### DISCUSSION TITLE
Can Authelia generate the Authorization header after succesful 2FA please?

### DISCUSSION THREAD

**Discussion by ArthurMitchell42** [authelia/authelia#6576] [Q&A] [answered]
I have Authelia set up with Traefik providing a very effective 2FA system to control access. I'm now writing a web app container using the Flask framework with Flask-HTTPAuth which expects the Authorization to be present in order to log the user into the frameworks ecosystem.
I see which auth headers Authelia generates but can't see a way to get the Authorization header from it to the backend service.
Advice would be appreciated.
A

**Comment by james-d-elliott** [comment, marked as answer]
There is no way currently however we'd like to implement a bearer scheme authorization header. If it expects basic that is something we just don't plan to implement unless the passwords for the backend systems are generated by Authelia.

Saving user passwords in plain text for re-authentication with a third party is one of the worst security decisions we can make. The only reasonable compromise here is if we can generate them and stipulate that they should not be given to an end user. 

Instead I'd probably suggest something like: this: https://flask-login.readthedocs.io/en/latest/#custom-login-using-request-loader

Specifically:

```python
@login_manager.request_loader
def load_user_from_request(request):
    username = request.headers.get('Remote-User')
    user = User.query.filter_by(username).first()
    if user:
        return user

    return None
```

**Reply by ArthurMitchell42** [reply]
Thanks James, can I assume that verifying the `host` attribute address against a trusted proxy setting in the app is also sensible? Also, is there any extra gained by checking the presence of the Authelia session cookie or any other header data please?
flask_httpauth does support bearer tokens, from their documents:

#### Token Authentication Example
The HTTPTokenAuth is a generic authentication handler that can be used with non-standard authentication schemes, with the scheme name given as an argument in the constructor. In the above example, the WWW-Authenticate header provided by the server will use Bearer as scheme:

WWW-Authenticate: Bearer realm="Authentication Required"

The verify_token callback receives the authentication credentials provided by the client on the Authorization header.

Rgds, A

**Reply by james-d-elliott** [reply]
Not sure what you mean by the host attribute address part. Could you elaborate?

You can realistically use the session cookie to verify it against the API however we do that with every request. I'll leave that up to you.

I can probably drum up an equivalent with Flask-HTTPAuth. It seems pretty customizable. We have aims to allow including the bearer token but yeah we'd need the backend to verify it. The only advantage here is that it would work with or without the proxy. As long as you ensure your proxy doesn't allow the header to be set at the user side there is no downside to using the `Remote-User` header to "login" the user though.

**Reply by ArthurMitchell42** [reply]
On the host attribute of the request object in Flask I was assuming it's good practice to verify the calling/redirecting host is the expected one (a trusted proxy setting in the app) to combat header spoofing, even if it's within the network?

**Reply by james-d-elliott** [reply]
If you wanted to verify anything it'd be the source IP address (TCP) of the proxy.

**Reply by ArthurMitchell42** [reply]
Evening,
Yes, thats what the host attribute gives you, the proxy machine. As I'm sure you know so well, the X-Forwarded-From has the original remote IP from Traefik

**Reply by ArthurMitchell42** [reply]
Thanks for your advice by the way, if you find the time to look at the Authorization->bearer method then do let me know and I'll try to support testing.
